README.md
# Zway-BlindControl

Manages blinds and shades for optimal shading when hot, and insulation when
cold. A virtual device will be created for each selected operation mode,
that can disable blind control actions.

It is possible to create multiple rules that manage several devices (eg. in
different rooms). Rules can be created based on

* Inside temperature
* Outside temperature
* Solar intensity (UV Index)
* Solar altitude
* Solar azimuth

The module supports two operation modes: Shading and insulation. For shading
blinds will be closed once all preconditions (in- and outside temperature and
solar intensity) match, and the sun shines into a particular window
(calculated based on altitude and azimuth). The blinds will stay closed until
the sun moves out of the selected area (regardless of the temperature) -
either because it set below the selected altitude, or because it does not
match the selected azimuth angles anymore.

In insulation mode the blinds will be closed if the temperature and the
altitude fall below a certain level. Blinds will be reopened once the solar
altitude rises above the selected level.

# Configuration

This module depends on the Astronomy module from
https://github.com/maros/Zway-Astronomy for calculating the solar azimuth
ana altitude.

## shade_active

Configures if shading management should be activated.

## shade_rules

A list of all rules/devices that should be managed. All selected conditions
must match for the blinds to be closed.

## shade_rules.temperature_outside

If the outside temperature rises above the selected level, this rule will
be activated (optional)

## shade_rules.temperature_inside

If the inside temperature rises above the selected level, this rule will
be activated (optional)

## shade_rules.sun_uv

If the UV index rises above the selected level, this rule will
be activated (optional).

## shade_rules.azimuth_left, shade_rules.azimuth_right

Select two azimuth levels (0-360°). Rule will be activated if sun is between
these two angles.

## shade_rules.altitude

Activate rule if sun rises above the selected altitude. (0-90°)

## shade_rules.devices

A list of all blinds and shades managed by this rule

## shade_rules.position

Desired closing position (0-99)

## insulation_active

Sets if low-temperature shade insulation mode should be activated.

## insulation_rules

A list of all rules/devices that should be managed. All selected conditions
must match for the blinds to be closed.

## insulation_rules.temperature_outside

If the outside temperature falls below the selected level, this rule will
be activated (optional)

## insulation_rules.altitude

Activate rule if sun falls below the selected altitude.

## insulation_rules.devices

A list of all blinds and shades managed by this rule.

## insulation_rules.position

Desired closing position (0-99)

## uv_sensor

Device that measures the UV index. Can be either a physical device (eg. Aeon
Labs Multisensor Gen 5) or a virtual device created by the WeatherUnderground
module ( https://github.com/maros/Zway-WeatherUnderground )

## temperature_outside_sensor

Device that measures the outside temperature. Can be either a physical device
or a virtual device created by any weather module
(eg. https://github.com/maros/Zway-WeatherUnderground )

## temperature_inside_sensor

Device that measures the inside temperature. If you want to use the average
temperature of multiple sensors, you can use the SummarySensor module
( https://github.com/maros/Zway-SummarySensor )

# Virtual Devices

This module creates two virtual binary that enables/disabled the shade and
insulation behaviour. The devices will only be created if the respective
rules are enabled in the configuration.

# Events

No events are emitted.

Blind controller always listens to security.smoke.alarm and
security.smoke.cancel events (these events are usually emitted by
( https://github.com/maros/Zway-SecurityZone ). In case of a smoke alarm all
blinds managed by the controller are automatically opened.

# Installation

Make sure that the Astronomy module and BaseModule is installed prior to
installing this module ( https://github.com/maros/Zway-Astronomy and
https://github.com/maros/Zway-BaseModule)

The prefered way of installing this module is via the "Zwave.me App Store"
available in 2.2.0 and higher. For stable module releases no access token is
required. If you want to test the latest pre-releases use 'k1_beta' as
app store access token.

For developers and users of older Zway versions installation via git is
recommended.

```shell
cd /opt/z-way-server/automation/userModules
git clone https://github.com/maros/Zway-BlindControl.git BlindControl --branch latest
```

To update or install a specific version
```shell
cd /opt/z-way-server/automation/userModules/BlindControl
git fetch --tags
# For latest released version
git checkout tags/latest
# For a specific version
git checkout tags/1.02
# For development version
git checkout -b master --track origin/master
```

# License

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or any
later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.


CHANGES
1.08 2019 Apr 26, Fri
    * metrics:target may be null
    * Use device target level as current position
    * Clarify log messages
    * Update copyright

1.07 2016 Dec 05, Mon
    * Fix winter mode calculations

1.06 2016 May 31, Tue
    * Add translation
    * Also accept solar intensity device
    * Fix azimuth comparison
    * One temperature sensor per zone
    * Add verbose logging
    * Improve texts
    * Update module title and description
    * Fix wrong variable name
    * Set defaults
    * Do not close blinds shortly before period end
    * Update probeType

1.05 2016 Apr 20, Wed
    * Fix time fields in module.json
    * Update readme
    * Set probeType
    * Cleanup device switching
    * More logging
    * Translation
    * Fix icon path
    * Use imagePath from BaseModule

1.04 2016 Mar 08, Tue
    * Fix wrong variable
    * Fix event names
    * Use helper to loop all devices
    * Listen to stop smoke event
    * Update installation howto
    * Cleanup modes
    * Clarify variable name

1.03 2016 Jan 22, Fri
    * Use forecast for blind shade
    * Improve german translation
    * Update config helper
    * Basic german translation
    * Add doc link
    * Add forecast temp devices
    * Use time instead of solar altitude to handle insulation
    * Remember active shade/insulation rules

1.02 2016 Jan 08, Fri
    * Rename title/description key to make module work in appstore Update readme
    * Update documentation
    * Use BaseModule
    * Unify variable names
    * Read probeType
    * set probe type instead of title
    * Reverse auto flag
    * Fix blind position confusion
    * Check for correct auto device labels
    * Update category

1.01 2015 Nov 15, Sun
    * Update documentation
    * Cleanups
    * Update astronomy probeTitle
    * Max 255 for multilevel
    * Improve alarm handling

1.00 2015 Nov 04, Wed
    * Initial Release 


index.js
/*** BlindControl Z-Way HA module *******************************************

Version: 1.08
(c) Maroš Kollár, 2015-2017
-----------------------------------------------------------------------------
Author: Maroš Kollár <maros@k-1.com>
Description:
    Manage blinds for optimal insulation and shading

******************************************************************************/

function BlindControl (id, controller) {
    // Call superconstructor first (AutomationModule)
    BlindControl.super_.call(this, id, controller);

    this.modes              = ['shade','insulation'];
    _.each(this.modes,function(type) {
        this[type+'Device'] = undefined;
    });

    this.insulationDevice   = undefined;
    this.interval           = undefined;
    this.sunDevice          = undefined;
    this.alarmCallback      = undefined;
    this.allDevices         = [];
}

inherits(BlindControl, BaseModule);

_module = BlindControl;

// ----------------------------------------------------------------------------
// --- Module instance initialized
// ----------------------------------------------------------------------------

BlindControl.prototype.init = function (config) {
    BlindControl.super_.prototype.init.call(this, config);

    var self = this;

    // Create vdev
    _.each(self.modes,function(type) {
        if (config[type+'Active'] === true) {
            self[type+'Device'] = this.controller.devices.create({
                deviceId: "BlindControl_"+type+'_'+ self.id,
                defaults: {
                    metrics: {
                        active: [],
                        title: self.langFile[type+'_active_label'],
                        level: 'off',
                        icon: self.imagePath+'/icon_'+type+'_off.png'
                    }
                },
                handler: _.bind(self.commandDevice,self,type),
                overlay: {
                    probeType: 'controller_blind_'+type,
                    deviceType: 'switchBinary'
                },
                moduleId: self.id
            });
        }
    });

    self.alarmCallback = _.bind(self.processAlarm,self);
    self.controller.on('security.smoke.alarm',self.alarmCallback);
    self.controller.on('security.smoke.stop',self.alarmCallback);

    self.interval = setInterval(_.bind(self.checkConditions,self),1000*60*3);
    setTimeout(_.bind(self.initCallback,self),1000*60);
};

BlindControl.prototype.stop = function () {
    var self = this;

    _.each(self.modes,function(type) {
        var key = type+'Device';
        if (typeof(self[key]) !== 'undefined') {
            self.controller.devices.remove(self[key].id);
            self[key] = undefined;
        }
    });

    if (typeof(self.interval) !== 'undefined') {
        clearInterval(self.interval);
        self.interval = undefined;
    }

    self.controller.off('security.smoke.alarm',self.alarmCallback);
    self.controller.off('security.smoke.stop',self.alarmCallback);

    self.alarmCallback = undefined;

    BlindControl.super_.prototype.stop.call(this);
};

BlindControl.prototype.initCallback = function() {
    var self = this;

    var devices = [];
    _.each(self.config.insulationRules,function(rule) {
        devices.push(rule.devices);
    });
    _.each(self.config.shadeRules,function(rule) {
        devices.push(rule.devices);
    });

    self.allDevices = _.uniq(_.flatten(devices));

    self.processDeviceList(self.allDevices,function(deviceObject) {
        var deviceAuto  = deviceObject.get('metrics:auto');
        if (typeof(deviceAuto) === 'undefined') {
            deviceObject.set('metrics:auto',false);
        } else if (deviceAuto === true) {
            var deviceLevel = deviceObject.get('metrics:level');
            if (deviceLevel > 99) {
                deviceObject.set('metrics:auto',false);
            }
        }
    });
};

// ----------------------------------------------------------------------------
// --- Module methods
// ----------------------------------------------------------------------------

BlindControl.prototype.commandDevice = function(type,command) {
    var self = this;

    if (command !== 'on' && command !== 'off') return;

    self.log('Turn '+type+' blind control '+command);
    var device = self[type+'Device'];
    device.set('metrics:level',command);
    device.set('metrics:icon',self.imagePath+'/icon_'+type+'_'+command+'.png');

    if (command === 'on') {
        var otherType = (type === 'shade') ? 'insulation':'shade';
        var otherDevice = self[otherType+'Device'];
        if (typeof(otherDevice) !== 'undefined') {
            otherDevice.performCommand('off');
        }
    } else if (command === 'off') {
        var active = device.get('metrics:active',command);
        // TODO re-open all blinds?
        _.each(self.config[type+'Rules'],function(rule,ruleIndex) {
            if (active[ruleIndex] === true) {
                self.moveDevices(rule.devices,255);
            }
        });
        self[type+'Device'].set('metrics:active',[]);
    }
};

BlindControl.prototype.checkConditions = function() {
    var self = this;

    self.log('Evaluating blind positions');
    if (self.config.insulationActive === true &&
        self.insulationDevice.get('metrics:level') === 'on') {
        self.processInsulationRules();
    }
    if (self.config.shadeActive === true &&
        self.shadeDevice.get('metrics:level') === 'on') {
        self.processShadeRules();
    }
};

BlindControl.prototype.processInsulationRules = function() {
    var self = this;

    var rulesActive         = self.insulationDevice.get('metrics:active');
    var outsideTemperature  = self.getSensorData(self.config.temperatureOutsideSensor);
    var foreacastLow        = self.getSensorData(self.config.forecastLowSensor) || outsideTemperature;
    var temperature         = Math.min(outsideTemperature,foreacastLow);

    if (typeof(temperature) === 'undefined') {
        self.error('Could not find outside temperature or forecast sensor');
        return;
    }

    _.each(self.config.insulationRules,function(rule,ruleIndex) {
        var isActive        = rulesActive[ruleIndex] || false;
        var timeFrom        = self.parseTime(rule.timeFrom);
        var timeTo          = self.parseTime(rule.timeTo);
        var inPeriod        = self.checkPeriod(timeFrom,timeTo);
        var timeLimit       = new Date(timeFrom.getTime() + (timeFrom.getTime() - timeTo.getTime()) / 2);
        var inPeriodStart   = self.checkPeriod(timeFrom,timeLimit);
        var inTemperature   = temperature < rule.temperatureOutside;

        // Check sun altitude & temp
        if (inTemperature
            && inPeriod === true
            && isActive === false
            && inPeriodStart === true ) {
            self.log('Close blind for insulation');
            rulesActive[ruleIndex] = true;
            self.moveDevices(rule.devices,rule.position);
        } else if (inPeriod === false
            && isActive === true) {
            self.log('Open blind after insulation');
            rulesActive[ruleIndex] = false;
            self.moveDevices(rule.devices,255);
        } else {
            self.log('Nothing to do. temp: '+inTemperature+' active:'+isActive+' period:'+inPeriod+' ('+inPeriodStart+')');
            return;
        }

        self.insulationDevice.set('metrics:active',rulesActive);
    });
};

BlindControl.prototype.processShadeRules = function() {
    var self = this;

    var rulesActive         = self.shadeDevice.get('metrics:active');
    var sunAltitude         = self.getSunAltitude();
    var sunAzimuth          = self.getSunAzimuth();
    var outsideTemperature  = self.getSensorData(self.config.temperatureOutsideSensor);
    var forecastHigh        = self.getSensorData(self.config.forecastHighSensor) || outsideTemperature;
    var uvIndex             = self.getSensorData(self.config.uvSensor);
    var temperature         = Math.max(outsideTemperature,forecastHigh);

    if (typeof(temperature) === 'undefined') {
        self.error('Could not find outside temperature or forecast sensor');
        return;
    }

    _.each(self.config.shadeRules,function(rule,ruleIndex) {
        var matchClose          = true;
        var matchPosition       = true;
        var isActive            = rulesActive[ruleIndex] || false;
        var insideTemperature   = self.getSensorData(rule.temperatureInsideSensor);

        // Check outside temperature
        if (temperature < rule.temperatureOutside) {
            self.log('Zone '+ruleIndex+'. Outside temperature too low  ('+temperature+'). Not closing');
            matchClose = false;
        }

        // Check inside temperature
        if (typeof(rule.temperatureInside) !== 'undefined') {
            if (typeof(insideTemperature) === 'undefined') {
                self.error('Could not find inside temperature sensor');
                return;
            }
            if (insideTemperature < rule.temperatureInside) {
                self.log('Zone '+ruleIndex+'. Inside temperature too low ('+insideTemperature+'). Not closing');
                matchClose = false;
            }
        }

        // Check UV
        if (typeof(rule.sunUv) !== 'undefined') {
            if (typeof(uvIndex) === 'undefined') {
                self.error('Zone '+ruleIndex+'. Could not find UV sensor');
                return;
            }
            if (uvIndex < rule.sunUv) {
                self.log('Zone '+ruleIndex+'. UV Index too low ('+uvIndex+'). Not closing');
                matchClose = false;
            }
        }

        // Check solar altitude
        if (sunAltitude < rule.altitude) {
            self.log('Zone '+ruleIndex+'. Sun altitude too low ('+sunAltitude+'). Not closing');
            matchPosition = false;
        } else if (!
                (
                    (
                        (rule.azimuthLeft > rule.azimuthRight) &&
                        (sunAzimuth > rule.azimuthLeft || sunAzimuth < rule.azimuthRight)
                    )
                    ||
                    (
                        rule.azimuthLeft < rule.azimuthRight &&
                        sunAzimuth > rule.azimuthLeft &&
                        sunAzimuth < rule.azimuthRight
                    )
                )
            ) {
            self.log('Zone '+ruleIndex+'. Sun azimuth autside of range ('+sunAzimuth+'). Not closing');
            matchPosition = false;
        }

        // Check sun altitude
        if (matchClose === true
            && matchPosition === true
            && isActive === false) {
            // Close
            self.log('Zone '+ruleIndex+'. Closing');
            rulesActive[ruleIndex] = true;
            self.moveDevices(rule.devices,rule.position);
        } else if (matchPosition === false
            && isActive === true) {
            // Re-open
            self.log('Zone '+ruleIndex+'. Opening');
            rulesActive[ruleIndex] = false;
            self.moveDevices(rule.devices,255);
        } else {
            self.log('Zone '+ruleIndex+'. Nothing to do. active:'+isActive+' position:'+matchPosition+' close:'+matchClose);
            return;
        }
        self.shadeDevice.set('metrics:active',rulesActive);
    });
};

BlindControl.prototype.processAlarm = function(event) {
    var self = this;

    var alarmed = true;

    console.logJS(event);
    // TODO check event type
    self.controller.devices.each(function(vDev) {
        if (vDev.get('probeType') === 'controller_security'
            && vDev.get('metrics:securityType') === 'smoke') {
            var state = vDev.get('metrics:state');
            if (state !== 'alarm' && state !== 'timeout') {
                alarmed = false;
            }
        }
    });

    if (alarmed) {
        self.log('Opening all blinds due to smoke alarm');
    }

    self.processDeviceList(self.allDevices,function(deviceObject) {
        if (alarmed) {
            deviceObject.set('metrics:auto',true);
            deviceObject.performCommand('on');
        } else {
            deviceObject.set('metrics:auto',false);
        }
    });
};

BlindControl.prototype.moveDevices = function(devices,targetPos) {
    var self = this;

    self.log('Move devices to '+targetPos);

    self.processDeviceList(devices,function(deviceObject) {
        var deviceAuto  = deviceObject.get('metrics:auto');
        var devicePos   = deviceObject.get('metrics:target');
        if (typeof(devicePos) === 'undefined' || devicePos === null) {
            devicePos       = deviceObject.get('metrics:level');
        }

        // Open
        if (targetPos >= 99 && (devicePos >= 99 || deviceAuto === false)) {
            self.log('Ignoring device open. Already at '+devicePos+' or not auto');
            if (deviceAuto === true) {
                deviceObject.set('metrics:auto',false);
            }
            return;
        // Close
        } else if  (targetPos < 99 && (devicePos < 99 || deviceAuto === true)) {
            self.log('Ignoring device close. Already at '+devicePos+' or auto');
            return;
        }

        self.log('Auto move blind '+deviceObject.id+' to '+targetPos);
        if (targetPos === 0) {
            deviceObject.set('metrics:auto',true);
            deviceObject.performCommand('off');
        } else if (targetPos >= 99) {
            deviceObject.set('metrics:auto',false);
            deviceObject.performCommand('on');
        } else {
            deviceObject.set('metrics:auto',true);
            deviceObject.performCommand('exact',{ level: targetPos });
        }
    });
};

BlindControl.prototype.getSunAzimuth = function() {
    var self = this;
    return self.getDeviceValue([
        ['probeType','=','astronomy_sun_altitude']
    ],'metrics:azimuth');
};

BlindControl.prototype.getSunAltitude = function() {
    var self = this;
    return self.getDeviceValue([
        ['probeType','=','astronomy_sun_altitude']
    ],'metrics:altitude');
};

BlindControl.prototype.getSensorData = function(deviceId) {
    var self = this;

    if (typeof(deviceId) === 'undefined') {
        return;
    }
    var deviceObject = self.controller.devices.get(deviceId);
    if (deviceObject === null) {
        return;
    }
    return deviceObject.get('metrics:level');
};